File: app/src/main/java/com/example/githubrepo/RepoListAdapter.java
package com.example.githubrepo;

import android.content.Context;
import android.content.Intent;
import android.graphics.Typeface;
import android.net.Uri;
import android.text.SpannableString;
import android.text.format.DateUtils;
import android.text.style.StyleSpan;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.example.githubrepo.databinding.ListItemRepoBinding;
import com.example.githubrepo.models.Repository;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.List;

/**
 * Created by sarahneo on 20/2/17.
 */

public class RepoListAdapter extends RecyclerView.Adapter<RepoListAdapter.ViewHolder> {

    private final LayoutInflater inflater;
    private final AppCompatActivity mActivity;
    private final List<Repository> mRepoList;

    public RepoListAdapter(AppCompatActivity activity, List<Repository> repoList) {
        this.mActivity = activity;
        this.mRepoList = repoList;
        inflater = (LayoutInflater) mActivity.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
    }

    @NonNull
    @Override
    public RepoListAdapter.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        ListItemRepoBinding binding = ListItemRepoBinding.inflate(inflater, parent, false);

        return new ViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(final RepoListAdapter.ViewHolder holder, int position) {

        final Repository item = mRepoList.get(position);

        String fullName = item.getFullName();
        SpannableString spanString = new SpannableString(fullName);
        spanString.setSpan(new StyleSpan(Typeface.BOLD), fullName.indexOf("/") + 1, spanString.length(), 0);
        spanString.setSpan(new StyleSpan(Typeface.ITALIC), fullName.indexOf("/") + 1, spanString.length(), 0);
        holder.mBinding.tvFullName.setText(spanString);

        holder.mBinding.tvDesc.setText(item.getDesc());
        holder.mBinding.tvLanguage.setText(item.getLanguage());
        holder.mBinding.tvStars.setText(Integer.toString(item.getStargazersCount()));
        holder.mBinding.tvForks.setText(Integer.toString(item.getForksCount()));
        if (item.getPushedAt() != null) {
            holder.mBinding.tvLastUpdated.setText(DateUtils.getRelativeTimeSpanString(
                    convertStringToTimeMillis(item.getPushedAt()),
                    System.currentTimeMillis(),
                    DateUtils.DAY_IN_MILLIS));
        }

        Glide.with(mActivity)
                .load(item.getOwnerAvatarUrl())
                .diskCacheStrategy(DiskCacheStrategy.RESULT)
                .into(holder.mBinding.civProfile);
    }

    @Override
    public int getItemCount() {
        return mRepoList.size();
    }

    class ViewHolder extends RecyclerView.ViewHolder {
        ListItemRepoBinding mBinding;

        ViewHolder(ListItemRepoBinding binding) {
            super(binding.getRoot());
            mBinding = binding;
            mBinding.getRoot().setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    Repository repo = mRepoList.get(getAdapterPosition());
                    String url = repo.getHtmlUrl();
                    Intent i = new Intent(Intent.ACTION_VIEW);
                    i.setData(Uri.parse(url));
                    if (i.resolveActivity(mActivity.getPackageManager()) != null)
                        mActivity.startActivity(i);
                }
            });
        }
    }

    private static long convertStringToTimeMillis(String dateString) {
        long milliseconds = 0;
        SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssZ");
        try {
            Date d = sdf.parse(dateString.replaceAll("Z$", "+0000"));
            milliseconds = d.getTime();
        } catch (ParseException e) {
            e.printStackTrace();
        }

        return milliseconds;
    }
}

File: app/src/main/java/com/example/githubrepo/RepoListAdapter.kt
package com.example.githubrepo

import android.content.ActivityNotFoundException
import android.content.Context
import android.content.Intent
import android.graphics.Typeface
import android.net.Uri
import android.text.SpannableString
import android.text.format.DateUtils
import android.text.style.StyleSpan
import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.Toast
import android.widget.Toast.LENGTH_SHORT
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.example.githubrepo.databinding.ListItemRepoBinding
import com.example.githubrepo.models.Repository
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.Locale

class RepoListAdapter(
        private val activity: AppCompatActivity,
        private val repoList: List<Repository>
) : RecyclerView.Adapter<RepoListAdapter.ViewHolder>() {
    private val inflater = activity.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val binding: ListItemRepoBinding = ListItemRepoBinding.inflate(inflater, parent, false)
        return ViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val item = repoList[position]
        val fullName = item.fullName.orEmpty()
        val spanString = if (fullName.isEmpty()) {
            SpannableString("")
        } else {
            SpannableString(fullName).also {
                it.setSpan(StyleSpan(Typeface.BOLD), fullName.indexOf("/") + 1, it.length, 0)
                it.setSpan(StyleSpan(Typeface.ITALIC), fullName.indexOf("/") + 1, it.length, 0)
            }
        }
        with(holder.binding) {
            tvFullName.text = spanString
            tvDesc.text = item.desc
            tvLanguage.text = item.language
            tvStars.text = item.stargazersCount.toString()
            tvForks.text = item.forksCount.toString()
        }
        item.pushedAt?.let {
            holder.binding.tvLastUpdated.text = DateUtils.getRelativeTimeSpanString(
                    convertStringToTimeMillis(it),
                    System.currentTimeMillis(),
                    DateUtils.DAY_IN_MILLIS)
        }
        Glide.with(activity)
                .load(item.ownerAvatarUrl)
                .diskCacheStrategy(DiskCacheStrategy.RESULT)
                .into(holder.binding.civProfile)
    }

    override fun getItemCount(): Int {
        return repoList.size
    }

    inner class ViewHolder(val binding: ListItemRepoBinding) : RecyclerView.ViewHolder(binding.root) {

        init {
            this.binding.root.setOnClickListener {
                val url = repoList[adapterPosition].htmlUrl
                val intent = Intent(Intent.ACTION_VIEW).also { it.data = Uri.parse(url) }
                try {
                    activity.startActivity(intent)
                } catch (e: ActivityNotFoundException) {
                    Toast.makeText(activity, "No browser app available", LENGTH_SHORT).show()
                }
            }
        }
    }

    private fun convertStringToTimeMillis(dateString: String): Long {
        var milliseconds: Long = 0
        val sdf = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssZ", Locale.US)
        try {
            val date = sdf.parse(dateString.replace("Z$".toRegex(), "+0000"))
            milliseconds = date?.time ?: 0
        } catch (e: ParseException) {
            e.printStackTrace()
        }
        return milliseconds
    }
}